### PR TITLE
(SERVER-2532) Enable caching for transports

### PR DESCRIPTION
**Pull request by justinstoller**
This depends on https://github.com/puppetlabs/puppetserver/pull/2065 see https://github.com/justinstoller/puppetserver/compare/server2471-3...justinstoller:server2532 for a realistic diff.

I need to update the tests at a minimum. Though I'd love to go through and rename all our caching, but I'm not sure if that's just a nice to have, that could potentially do later.

### CODE REVIEW

File: src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj [hunk 1]
```diff
@@ -99,6 +99,12 @@
                                  (tk-services/service-context this))]
      (get-in @environment-class-info [env-name :info :classes])))
 
+  (get-environment-info-tag
```
Comment (Magisus) [inline src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj]: Did you leave the `class`-specific version of this function on purpose? Are we required to maintain the protocol or something? Because I think this commit gets rid of the only place it's called...

File: src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj [hunk 2]
```diff
@@ -375,6 +375,30 @@
         (update-environment-info-entry cache-entry tag))
       environment-info-cache)))
 
+(schema/defn ^:always-validate
+  environment-info-cache-updated-with-tag :- EnvironmentInfoCache
+  "Return the supplied environment class info cache argument, updated per
```
Comment (Magisus) [inline src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj]: This comment still refer to `class` several times.

File: src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj [hunk 3]
```diff
@@ -113,6 +123,17 @@
             tag
             cache-generation-id-before-tag-computed)))
 
+  (set-environment-info-tag!
```
Comment (Magisus) [inline src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj]: Same question here, are we obligated to leave the `class`-specific version?

File: test/integration/puppetlabs/services/master/environment_transports_int_test.clj [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+                "response: "
+                (ks/pprint-to-string response)))
+           (is (not= initial-etag (response-etag response))
+               "etag did not change even though code did")
+           (is (= mock-response
+                  (response->map response))
+               "unexpected body following code change")))))))
+
+
+#_(deftest ^:integration environment-transports-integration-cache-enabled-test3
```
Comment (Magisus) [inline test/integration/puppetlabs/services/master/environment_transports_int_test.clj]: Can this be enabled now?

File: test/integration/puppetlabs/services/master/environment_transports_int_test.clj [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+          (.add a t)
+          a))
+      (handleRequest [_ _]
+        (JRubyPuppetResponse. 0 nil nil nil))
+      (puppetVersion [_]
+        "1.2.3")
+      (terminate [_]
+        (log/info "Terminating Master")))))
+
+#_(deftest ^:integration transports-info-updated-after-cache-flush-during-prior-request
```
Comment (Magisus) [inline test/integration/puppetlabs/services/master/environment_transports_int_test.clj]: Can this be enabled now?

File: src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj (hunk 1 above)
Comment (justinstoller) [reply src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj]: So far this PR only changes names of those methods internal to jruby-puppet-core and then adds new functions where needed by external consumers. I'd love to remove all the old ones, replace them with generic ones, and remove "class" from any cache references, but that will add a lot of busy work that's orthogonal to the actual functionality and didn't know if I should do it here or in a follow on PR.

Comment (Magisus) [reply src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj]: Got it, this is what you were referring to in your top comment. Sure, it can wait for a follow-on.

**Comment by Magisus** [comment]
We should have docs for this endpoint similar to the others we've written recently. Don't have a strong opinion about whether that goes in with this PR or as its own thing.

**Comment by Magisus** [comment]
Manually verified that this worked in a PE install in combination with https://github.com/puppetlabs/puppet-enterprise-modules/pull/318 and the [panos module](https://forge.puppet.com/puppetlabs/panos).

**Comment by justinstoller** [comment]
I'll clean up the commits if folks want to otherwise review (I'm also happy to break out the last bit of work into separate PRs).

**Comment by Magisus** [comment]
Definitely a lot in here. But I think most of the refactors make sense. I don't know that the test refactors at the end need to be in a separate PR (though they could be), but I would appreciate a commit message about why you decided to switch over to using the mock JRuby.

File: src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj [hunk 6]
```diff
@@ -281,73 +281,51 @@
...
-  (update-environment-class-info-entry
-   original-environment-class-info-entry
-   nil))
+  {:classes {:tag nil :version 1}
+   :transports {:tag nil :version 1}})
+
+(schema/defn ^:always-valid invalidate-environment-info-entry
+  :- EnvironmentInfoCacheEntry
+  "Return the supplied 'original-environment-info-entry', only updated
+  with nil tags and a cache-generation-id value that has been incremented."
```
Comment (pcarlisle) [inline src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj]: `cache-generation-id` again

File: src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj [hunk 7]
```diff
@@ -360,54 +338,93 @@
...
-      (get-in [env-name :cache-generation-id])))
+      (get-in [env-name info-service :version])))
 
 (schema/defn ^:always-validate
   mark-environment-expired!
   "Mark the specified environment expired, in all JRuby instances.  Resets
-  the cached class info for the environment's 'tag' to nil and increments the
-  'cache-generation-id' value."
+  the cached tags for the environment's info services to nil and increments
+  the 'cache-generation-id' value."
```
Comment (pcarlisle) [inline src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj]: `cache-generation-id`, should probably just search for these

File: src/clj/puppetlabs/services/master/master_core.clj [hunk 8]
```diff
@@ -602,10 +533,23 @@
...
+  (let [path-components (-> request :route-info :path)
+        ;; path-components will be something like
+        ;; ["/puppet" "/v3" "/environment_classes" ["*" :rest]]
+        ;; and we want to map "/environment_classes" to a
+        ;; cacheable info service
+        resource-component (-> path-components butlast last)]
+    (when resource-component
+      (get {"environment_classes" :classes
+            "environment_transports" :transports}
+           (str/replace resource-component "/" "")))))
```
Comment (pcarlisle) [inline src/clj/puppetlabs/services/master/master_core.clj]: you could have just put the / in the strings in the map

File: src/clj/puppetlabs/services/master/master_core.clj [hunk 9]
```diff
@@ -614,23 +558,85 @@
...
+  "Given a retrieval fn of the jruby protocol, builds a handler for fn that
+  honors the environment cache."
+  [info-fn :- IFn
+   jruby-service :- (schema/protocol jruby-protocol/JRubyPuppetService)
+   cache-enabled? :- schema/Bool]
+  (fn [request]
+    (let [env (jruby-request/get-environment-from-request request)
+          svc-id (info-service request)
+          cache-id
+           (jruby-protocol/get-environment-cache-version!
```
Comment (pcarlisle) [inline src/clj/puppetlabs/services/master/master_core.clj]: this fits on the previous line

File: src/clj/puppetlabs/services/master/master_core.clj [hunk 10]
```diff
@@ -614,23 +558,85 @@
...
+         jruby-service
+         env
+         svc-key
+         tag
+         cache-version)
+        (-> (response-with-etag body tag)
+            (rr/content-type "application/json"))))))
+
+(schema/defn ^:always-validate
+  mk-cacheable-handler :- IFn
```
Comment (pcarlisle) [inline src/clj/puppetlabs/services/master/master_core.clj]: just write out "make"

File: test/integration/puppetlabs/services/master/environment_transports_int_test.clj [hunk 11]
```diff
@@ -0,0 +1,674 @@
...
+             @(:continue-promise promises))
+           a))
+       (handleRequest [_ _]
+         (JRubyPuppetResponse. 0 nil nil nil))
+       (puppetVersion [_]
+         "1.2.3")
+       (terminate [_]
+         (log/info "Terminating Master"))))))
+
+(defn mkmod
```
Comment (pcarlisle) [inline test/integration/puppetlabs/services/master/environment_transports_int_test.clj]: its esr 2 rd if u dnt abrv

File: src/clj/puppetlabs/services/master/master_core.clj [hunk 12]
```diff
@@ -614,23 +558,85 @@
...
-                              environment-class-cache-enabled)
+  raw-transports->response-map
+  [data :- List
+   env :- schema/Str]
+  (sort-nested-info-maps
+    {:name env
+     :transports data}))
+
+(schema/defn ^:always-validate
+  check-cache! :- ringutils/RingResponse
```
Comment (pcarlisle) [inline src/clj/puppetlabs/services/master/master_core.clj]: should be named closer to what it does, i'd expect `check-cache` to be used to check the cache early in the request to avoid doing work, but this is used after only for updating the cache

File: src/clj/puppetlabs/services/master/master_core.clj [hunk 13]
```diff
@@ -614,23 +558,85 @@
...
+  "Updates cache if new data does not match previous data
+
+  We must use the cache version generated at the start of the request.
+  The cache will check to see that the version given to it is the same
+  as its internal version (eg that it hasn't been reset since handing
+  out that version). If the given and internal versions do not match the
+  request to set the tag is ignored, if they match the tag is stored and
+  the internal version is incremented."
```
Comment (pcarlisle) [inline src/clj/puppetlabs/services/master/master_core.clj]: this whole paragraph could be more clear

File: test/integration/puppetlabs/services/master/environment_transports_int_test.clj (hunk 11 above)
Comment (justinstoller) [reply test/integration/puppetlabs/services/master/environment_transports_int_test.clj]: dont rag on my a e s t h e t i c, man.

File: src/clj/puppetlabs/services/master/master_core.clj (hunk 8 above)
Comment (justinstoller) [reply src/clj/puppetlabs/services/master/master_core.clj]: the slash seemed incidental to how we were doing the routing rather than the semantics of the route segment. it be weird for this to break if the segments changed from "/environment_classes" to "environment_classes" or "environment_classes/". but maybe that's yagni?

Comment (pcarlisle) [reply src/clj/puppetlabs/services/master/master_core.clj]: you can do it like this too it's not a big deal, just seemed like it added an extra call in here that you didn't need

Comment (pcarlisle) [reply src/clj/puppetlabs/services/master/master_core.clj]: i don't care that much either way

**Comment by justinstoller** [comment]
I think I've addressed the recent comments, if so I'll rebase.

**Comment by pcarlisle** [comment]
ready for rebase i think

**Comment by justinstoller** [comment]
rebased, I struggled writing the headline part of the commit, let me know if they should be reworked, or if I should split/combine any others.
